x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_save.ts

```diff
@@ -0,0 +1,297 @@
+                        : null,
+                  },
+                  ...(isMatchAllFilter(basicFilter)
+                    ? {
+                        query: {
+                          match_all: convertToString(
+                            (basicFilter as MatchAllFilter).query.match_all
+                          ),
+                        },
+                      }
+                    : { match_all: null }),
+                  ...(isExistsFilter(basicFilter) && basicFilter.query.exists != null
+                    ? { query: { exists: convertToString(basicFilter.query.exists) } }
+                    : { exists: null }),
+                  ...((isQueryStringFilter(basicFilter) || get('query', basicFilter) != null) &&
+                  basicFilter.query != null
+                    ? { query: convertToString(basicFilter.query) }
+                    : { query: null }),
+                  ...(isRangeFilter(basicFilter) && basicFilter.query.range != null
+                    ? { query: { range: convertToString(basicFilter.query.range) } }
+                    : { range: null }),
+                  ...(isScriptedRangeFilter(basicFilter) &&
+                  basicFilter.query.script !=
+                    null /* TODO remove it when PR50713 is merged || esFilters.isPhraseFilter(basicFilter) */
+                    ? { query: { script: convertToString(basicFilter.query.script) } }
+                    : { script: null }),
+                };
+              })
+            : [],
+          acc
+        );
+      }
+      return set(key, get(key, timeline), acc);
+    }
+    return acc;
+  }, timelineInput);
+
+const convertToString = (obj: unknown) => {
+  try {
+    if (isObject(obj)) {
+      return JSON.stringify(obj);
+    }
+    return fpToString(obj);
+  } catch {
+    return '';
+  }
+};
+
+type PossibleResponse = TimelineResponse | TimelineErrorResponse;
+
+function isTimelineErrorResponse(response: PossibleResponse): response is TimelineErrorResponse {
+  return 'status_code' in response || 'statusCode' in response;
+}
+
+function getErrorFromResponse(response: TimelineErrorResponse) {
+  if ('status_code' in response) {
+    return { errorCode: response.status_code, message: response.message };
+  } else if ('statusCode' in response) {
+    return { errorCode: response.statusCode, message: response.message };
+  }
+}
+
+function extractTimelineIdsAndVersions(timeline: TimelineModel) {
+  // When a timeline hasn't been saved yet, its `savedObectId` is not defined.
+  // In that case, we want to overwrite all locally created properties for the
+  // timeline id, the timeline template id and the timeline template version.
+  return {
+    timelineId: timeline.savedObjectId ?? null,
+    timelineVersion: timeline.version,
+    templateTimelineId: timeline.savedObjectId ? timeline.templateTimelineId : null,
+    templateTimelineVersion: timeline.savedObjectId ? timeline.templateTimelineVersion : null,
+  };
```

Comment:
This logic is basically what the global `myEpicTimelineId` was used for before. It caused a lot of issues due to its mutable nature. This immutable version of the logic is documented and only lives in a single place and is called from one single place.